### PR TITLE
Add schema support for tasks.json

### LINKED ISSUE #1098: Assist the user in creating task configuration
ATM is the user has no or an empty task configuration file (tasks.json), the quick command dialog will be empty, for tasks. 

peek 2018-01-22 10-02

In VS Code, a skeleton tasks.json is created and opened in an editor, in this situation. 

Maybe we could do something similar. 

Bonus: add a JSON schema for tasks.json, so that the editor knows about the file format's, can suggest completions, validate the content.

### PR DESCRIPTION
**Pull request by gorkem**
Provides configuration assistance for tasks.json.

Adds schema for task configurations. It enhances task-server to be able to collect information about task types.

fixes #1098

### CODE REVIEW

File: packages/task/src/browser/task-schema-updater.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            properties: {
+                tasks: {
+                    type: 'array',
+                    items: {
+                        ...deepClone(taskConfigurationSchema)
+                    }
+                }
+            }
+        };
+        const taskTypes = await this.taskService.getRegisteredTaskTypes();
```
Comment (akosyakov) [inline packages/task/src/browser/task-schema-updater.ts]: Could it be that task types are changed dynamically? For example by an activation of a new VS Code extension. In this case the schema should be updated.

File: packages/task/src/node/task-runner.ts [hunk 2]
```diff
@@ -63,4 +63,12 @@ export class TaskRunnerRegistry {
         const runner = this.runners.get(type);
         return runner ? runner : this.defaultRunner;
     }
+
```
Comment (akosyakov) [inline packages/task/src/node/task-runner.ts]: We probably need an event here, something like `onDidRegisterRunner`.

File: packages/task/src/node/task-runner.ts [hunk 3]
```diff
@@ -63,4 +63,12 @@ export class TaskRunnerRegistry {
...
+    getRunnerTypes(): string[] {
+        const types = ['shell', 'process'];
+        for (const type of this.runners.keys()) {
+            types.push(type);
+        }
+        return types;
```
Comment (azatsarynnyy) [inline packages/task/src/node/task-runner.ts]: just spread it
`return [...this.runners.keys()];`

File: packages/task/src/node/task-server.ts [hunk 4]
```diff
@@ -67,6 +67,10 @@ export class TaskServerImpl implements TaskServer {
         return taskInfo;
     }
 
+    async getRegisteredTaskTypes(): Promise<string[]> {
+        return Promise.resolve(this.runnerRegistry.getRunnerTypes());
```
Comment (azatsarynnyy) [inline packages/task/src/node/task-server.ts]: It's the async function, so its return value will be automatically wrapped in a resolved promise
`return this.runnerRegistry.getRunnerTypes();`

File: packages/task/src/browser/task-schema-updater.ts (hunk 1 above)
Comment (gorkem) [reply packages/task/src/browser/task-schema-updater.ts]: Runners is a concept that is unique to Theia. I do not think there is a way to update them dynamically.

Comment (akosyakov) [reply packages/task/src/browser/task-schema-updater.ts]: ok

File: packages/task/src/browser/task-schema-updater.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                    properties: {
+                        label: {
+                            type: 'string',
+                            description: 'A unique string that identifies the task that is also used as task\'s user interface label'
+                        },
+                        type: {
+                            type: 'string',
+                            enum: ['shell', 'process'],
+                            default: 'shell',
+                            description: 'Determines what type of process will be used to execute the task. only shell types will have output shown on the user interface'
```
Comment (kittaakos) [inline packages/task/src/browser/task-schema-updater.ts]: [minor]: `only` -> `Only`

**Pull request merged**